readthedocs/projects/tasks/builds.py

```diff
@@ -556,85 +537,30 @@ def update_build(self, state):
             log.exception('Unable to update build')
 
     def execute(self):
-        self.run_setup()
-        self.run_build()
-
-    def run_setup(self):
-        """
-        Run setup in a build environment.
-
-        1. Create a Docker container with the default image
-        2. Clone the repository's code and submodules
-        3. Save the `config` object into the database
-        4. Update VCS submodules
-        """
-        environment = self.data.environment_class(
-            project=self.data.project,
-            version=self.data.version,
-            build=self.data.build,
-            environment=self.get_vcs_env_vars(),
-            # Force the ``container_image`` to use one that has the latest
-            # ca-certificate package which is compatible with Lets Encrypt
-            container_image=settings.RTD_DOCKER_BUILD_SETTINGS['os']['ubuntu-20.04']
-        )
-
-        # Environment used for code checkout & initial configuration reading
-        with environment:
-            before_vcs.send(
-                sender=self.data.version,
-                environment=environment,
-            )
-
-            self.setup_vcs(environment)
-            self.data.config = load_yaml_config(version=self.data.version)
-            self.save_build_config()
-            self.update_vcs_submodules(environment)
-
-    def update_vcs_submodules(self, environment):
-        version_repo = self.get_vcs_repo(environment)
-        if version_repo.supports_submodules:
-            version_repo.update_submodules(self.data.config)
-
-    def run_build(self):
-        """Build the docs in an environment."""
-        self.data.build_env = self.data.environment_class(
-            project=self.data.project,
-            version=self.data.version,
-            config=self.data.config,
-            build=self.data.build,
-            environment=self.get_build_env_vars(),
+        # NOTE: what about doing:
+        #
+        # (triggered)
+        # - self.clone()
+        # - self.install()
+        # - self.build()
+        # - self.upload()
+        # (finished)
```

Comment:
This feels really clean. I think the big question is the seperate of concerns. I see how we're updating the build data, and we're doing that in the task instead of the builder -- I'd like to be a bit more clear on why that is.

Reply:
I originally wrote this because I thought we would have more things to do on each state, but in the end, it's just to update the build's status and call one method from `DocumentationBuilder` class. So, now I think that splitting them into methods may add more overhead than another thing.

I updated this chunk of code just with small comments for each "section" and I think that may be enough for now. In the future, if we add more extra logic here, we can split them. What do you think?